src/btree/bt_cursor.c

```diff
@@ -395,65 +482,45 @@ __wt_cursor_valid(
          * even after checking the insert list above, is correct, see the comment below for details.
          */
         upd = cbt->ins ? cbt->ins->upd : NULL;
-        break;
-    case BTREE_ROW:
-        /* The search function doesn't check for empty pages. */
-        if (page->entries == 0)
-            return (0);
-        /*
-         * In case of prepare conflict, the slot might not have a valid value, if the update in the
-         * insert list of a new page scanned is in prepared state.
-         */
-        WT_ASSERT(session, cbt->slot == UINT32_MAX || cbt->slot < page->entries);
-
-        /*
-         * See above: for row-store, no insert object can have the same key as an on-page object,
-         * we're done.
-         */
-        if (cbt->ins != NULL)
-            return (0);
+    }
 
-        /* Paranoia. */
-        WT_ASSERT(session, recno == WT_RECNO_OOB);
+    WT_RET(__cursor_valid_int_ondisk(cbt, NULL, upd, valid));
+    return (0);
+}
 
-        /*
-         * The key can be NULL only when we didn't find an exact match, copy the search found key
-         * into the temporary buffer for further use.
-         */
-        if (key == NULL) {
-            WT_RET(__wt_row_leaf_key(
-              session, cbt->ref->page, &cbt->ref->page->pg_row[cbt->slot], cbt->tmp, true));
-            key = cbt->tmp;
-        }
-        if (check_bounds) {
-            WT_RET(__btcur_bounds_contains_key(
-              session, &cbt->iface, key, WT_RECNO_OOB, &key_out_of_bounds, NULL));
-            /* The key value pair we were trying to return weren't within the given bounds. */
-            if (key_out_of_bounds)
-                return (0);
-        }
+/*
+ * __wt_cursor_valid --
+ *     Return if the cursor references an valid key/value pair.
+ */
+int
+__wt_cursor_valid(WT_CURSOR_BTREE *cbt, bool *valid, bool check_bounds)
+{
+    *valid = false;
 
-        /* Check for an update. */
-        upd = (page->modify != NULL && page->modify->mod_row_update != NULL) ?
-          page->modify->mod_row_update[cbt->slot] :
-          NULL;
-        break;
+    /* If the cursor is operating in key only mode we don't need to do any more work. */
+    if (F_ISSET(&cbt->iface, WT_CURSTD_KEY_ONLY)) {
```

Comment:
Oh and in theory this is a logic change as well but is safe for all existing uses as far as I can tell. It certainly cleans up a lot of bloat in `__wt_btcur_search`.